### ISSUE TITLE
Future of `DataArray.rename`

### ISSUE THREAD

**Issue by headtr1ck** [pydata/xarray#6704]
### What is your issue?

In https://github.com/pydata/xarray/pull/6665 the question came up what to do with `DataArray.rename` in light of the new index refactor.

To be consistent with `Dataset` we should introduce a

- `DataArray.rename_dims`
- `DataArray.rename_vars`
- `DataArray.rename`

Several open questions about the behavior (Similar things apply to `Dataset.rename{, _dims, _vars}`):

- [ ] Should `rename_dims` also rename indexes (dimension coordinates)?
- [ ] Should `rename_vars` also rename the DataArray?
- [ ] What to do if the `DataArray` has the same name as one of its coordinates?
- [ ] Should `rename` still rename everything (like it is now) or only the name (Possibly with some deprecation cycle)?

The current implementation of `DataArray.rename` is a bit inconsistent:

As stated by @max-sixty in https://github.com/pydata/xarray/issues/6665#issuecomment-1154368202_:
- rename operates on DataArray as described in https://github.com/pydata/xarray/pull/6665#issuecomment-1150810485.%C2%A0Generally I'm less keen on "different types have different semantics", and here a positional arg would mean a DataArray rename, and kwarg would mean var rename. But it does work locally to DataArray quite well.
- rename only exists on DataArrays for the name of the DataArray, and we use rename_vars & rename_dims for both DataArrays & Datasets. So Dataset.rename is soft-deprecated.

**Comment by max-sixty** [comment]
This is the sort of thing we can discuss briefly on our dev call — which you would be more than welcome to attend and participate in @headtr1ck.

It's at 8.30AM PT every second Wednesday, including this coming Wednesday. I think there's an issue about it (though am on a plane wifi connection so currently difficult to search...). Unfortunately I've been poor at attendance recently given my schedule, but I have been trying to rearrange things so I can start attending again.

**Comment by keewis** [comment]
it's #4001

**Comment by headtr1ck** [comment]
I feel honored thanks! But given that I will become a father any day now, I don't think I will be able to join regularly...

**Comment by max-sixty** [comment]
> I feel honored thanks! But given that I will become a father any day now, I don't think I will be able to join regularly...

Congrats on the forthcoming arrival!

Ofc, no stress if you can't make the meeting, and obv zero stress if you can't make lots of them.

**Comment by mathause** [comment]
There is also anther issue discussing this: #4825

**Comment by max-sixty** [comment]
As discussed on the call:
- We'd like to move to `.rename` and `.rename_vars` as separate methods on `DataArray`; the latter renaming coords.
- We would "soft-deprecate" `.rename` for renaming coords — i.e. remove documentation, but not give any warnings, at least for a while. This was a successful approach for `.drop`.
- `.rename_vars` is a slightly odd name given there are only coords. But it's consistent with `xr.Dataset`.

After the call, I realize that I didn't mention that `Dataset.rename` is a synonym for `Dataset.rename_vars`. Would be also soft deprecate `Dataset.rename`?

**Comment by shoyer** [comment]
`Dataset.rename` does both variables *and* dimensions. That seems useful in many cases. I think it also makes more sense than `Dataset.drop` does, given that variables and dimensions often use the same names -- whereas `drop` mixed up variable names and index values.

**Comment by benbovy** [comment]
> Should rename_dims also rename indexes (dimension coordinates)?

Yes, renaming dimensions and/or coordinates should be properly handled by indexes. Xarray's `Index` abstract class provides a `rename` method that may (should) be implemented in subclasses.

For example, `PandasIndex` and `PandasMultiIndex` both have a `dim` attribute that must match the dimension name (this attribute is required as in a near future it will be possible to set indexes for non-dimension coordinates).

`PandasIndex.rename` (`PandasMultiIndex.rename`) also rename the underlying pandas index (levels) according to the new coordinate name(s).

**Comment by benbovy** [comment]
> Should rename_dims also rename indexes (dimension coordinates)?

Sorry I think I misunderstood the question. If that means "should rename_dims also rename the (dimension) coordinates", then I think the answer should be no. With the explicit indexes refactor coordinates and indexes are less tightly coupled, i.e., in the mid/long term we want to drop the concept of a "dimension coordinate with an implicit index". So I think `rename_dims` should only rename dimensions (and update indexes metadata accordingly, cf. my previous comment).

**Comment by max-sixty** [comment]
Found this after hitting the issue (and realize I have commented a few times a year ago...)

Note that there's some overlap between this & #6238 & #4825. 

@benbovy 's comments at https://github.com/pydata/xarray/issues/6238#issuecomment-1039990176 & https://github.com/pydata/xarray/issues/4825#issuecomment-916974087 seem compelling to me...
